### PR TITLE
Save 64 bytes from most Linux/amd64 interpreter frames.

### PR DESCRIPTION
**Pull request by jaykrell**
Probably same on all 64bit architectures. half on 32bit.

before:
```
0000000000003ce0 <interp_exec_method_full>:
    3ce0:	55                   	push   %rbp
    3ce1:	48 89 e5             	mov    %rsp,%rbp
    3ce4:	41 57                	push   %r15
    3ce6:	41 56                	push   %r14
    3ce8:	41 55                	push   %r13
    3cea:	41 54                	push   %r12
    3cec:	49 89 f4             	mov    %rsi,%r12
    3cef:	53                   	push   %rbx
    3cf0:	48 89 fb             	mov    %rdi,%rbx
    3cf3:	48 81 ec 78 01 00 00 	sub    $0x178,%rsp
```

after:
```
0000000000003ce0 <interp_exec_method_full>:
    3ce0:	55                   	push   %rbp
    3ce1:	48 89 e5             	mov    %rsp,%rbp
    3ce4:	41 57                	push   %r15
    3ce6:	41 56                	push   %r14
    3ce8:	41 55                	push   %r13
    3cea:	41 54                	push   %r12
    3cec:	49 89 f4             	mov    %rsi,%r12
    3cef:	53                   	push   %rbx
    3cf0:	48 89 fb             	mov    %rdi,%rbx
    3cf3:	48 81 ec 38 01 00 00 	sub    $0x138,%rsp
```

It was 0x158 recently.

Some frames will be larger, i.e. dynamic + pinvoke.

Classic technique:
 pointer instead of struct
 initialized to null
 alloca it when needed and not null

A more detailed analysis might enable better, such as making
the parameter optional, and/or having the callee default it to its
local storage.

### CODE REVIEW

File: mono/mini/interp/interp.c [hunk 1]
```diff
@@ -3259,14 +3257,15 @@ interp_exec_method_full (InterpFrame *frame, ThreadContext *context, FrameClause
...
 				MonoMarshalSpec **mspecs;
-				MonoMethodPInvoke piinfo;
 				MonoMethod *m;
 
 				/* Pinvoke call is missing the wrapper. See mono_get_native_calli_wrapper */
 				mspecs = g_new0 (MonoMarshalSpec*, csignature->param_count + 1);
-				memset (&piinfo, 0, sizeof (piinfo));
 
-				m = mono_marshal_get_native_func_wrapper (m_class_get_image (imethod->method->klass), csignature, &piinfo, mspecs, code);
+				piinfo = piinfo ? piinfo : g_newa (MonoMethodPInvoke, 1);
```
Comment (filipnavara) [inline mono/mini/interp/interp.c]: Where is it `free`d?

Comment (jaykrell) [reply mono/mini/interp/interp.c]: `a` is for `alloca`. When the function returns.

Comment (lewurm) [reply mono/mini/interp/interp.c]: https://github.com/mono/mono/blob/a6e6c72c8fc3702db37e1572694a49902f63195f/mono/eglib/glib.h#L292

probably not the best name.

Comment (filipnavara) [reply mono/mini/interp/interp.c]: Oh, I see.

Comment (jaykrell) [reply mono/mini/interp/interp.c]: It is subtle I agree.
Should I just spell out `alloca` and the cast?
`g_allocate_stack`? :)

Comment (filipnavara) [reply mono/mini/interp/interp.c]: It's fine I guess. The name `g_newa` is confusing since it doesn't imply it's `alloca` under the covers.

Comment (vargaz) [reply mono/mini/interp/interp.c]: Lets hope this is not called in a loop.

Comment (lewurm) [reply mono/mini/interp/interp.c]: @vargaz why would that matter?

Comment (vargaz) [reply mono/mini/interp/interp.c]: alloca cannot be called in a loop because it will cause a stack overflow.

Comment (lewurm) [reply mono/mini/interp/interp.c]: `piinfo` wouldn't be `NULL` anymore.

Comment (filipnavara) [reply mono/mini/interp/interp.c]: It is guarded to do the allocation only once so it would not matter here.

**Pull request merged**